app/src/main/java/com/google/android/sambadocumentsprovider/browsing/broadcast/BroadcastUtils.java

```diff
@@ -0,0 +1,179 @@
+   * https://tools.ietf.org/html/rfc1002
+   * Section 4.2.12
+   */
+  static byte[] createPacket(int transId) {
+    ByteBuffer os = ByteBuffer.allocate(50);
+
+    char broadcastFlag = 0x0010;
+    char questionCount = 1;
+    char answerResourceCount = 0;
+    char authorityResourceCount = 0;
+    char additionalResourceCount = 0;
+
+    os.putChar((char) transId);
+    os.putChar(broadcastFlag);
+    os.putChar(questionCount);
+    os.putChar(answerResourceCount);
+    os.putChar(authorityResourceCount);
+    os.putChar(additionalResourceCount);
+
+    // Length of name. 16 bytes of name encoded to 32 bytes.
+    os.put((byte) 0x20);
+
+    // '*' character encodes to 2 bytes.
+    os.put((byte) 0x43);
+    os.put((byte) 0x4b);
+
+    // Write the remaining 15 nulls which encode to 30* 0x41
+    for (int i = 0; i < 30; i++) {
+      os.put((byte) 0x41);
+    }
+
+    // Length of next segment.
+    os.put((byte) 0);
+
+    // Question type: Node status
+    os.putChar((char) 0x21);
+
+    // Question class: Internet
+    os.putChar((char) 0x01);
+
+    return os.array();
+  }
+
+  /**
+   * Parses a positive response to NetBIOS name request query.
+   * https://tools.ietf.org/html/rfc1002
+   * Section 4.2.13
+   */
+  static List<String> extractServers(byte[] data, int expectedTransId) throws BrowsingException {
+    List<String> servers = new ArrayList<>();
+
+    try {
+      ByteBuffer buffer = ByteBuffer.wrap(data).order(ByteOrder.BIG_ENDIAN);
+
+      int transId = buffer.getChar();
+      if (transId != expectedTransId) {
+        // This response is not to our broadcast.
+
+        if (BuildConfig.DEBUG) Log.d(TAG, "Irrelevant broadcast response");
+
+        return servers;
+      }
+
+      skipBytes(buffer, 2); // Skip flags.
+
+      skipBytes(buffer, 2); // No questions.
+      skipBytes(buffer, 2); // Skip answers count.
+      skipBytes(buffer, 2); // No authority resources.
+      skipBytes(buffer, 2); // No additional resources.
+
+      int nameLength = buffer.get();
+      skipBytes(buffer, nameLength);
+
+      skipBytes(buffer, 1);
+
+      int nodeStatus = buffer.getChar();
+      if (nodeStatus != 0x20 && nodeStatus != 0x21) {
+        throw new BrowsingException("Received negative response for the broadcast");
+      }
+
+      skipBytes(buffer, 2);
+      skipBytes(buffer, 4);
+      skipBytes(buffer, 2);
+
+      int addressListEntryCount = buffer.get();
+
+      for (int i = 0; i < addressListEntryCount; i++) {
+        byte[] nameArray = new byte[SERVER_NAME_LENGTH];
+        buffer.get(nameArray, 0, SERVER_NAME_LENGTH);
+
+        final String serverName = new String(nameArray, Charset.forName(SERVER_NAME_CHARSET));
+        final int type = buffer.get();
+
+        if (type == FILE_SERVER_NODE_TYPE) {
+          servers.add(serverName);
+        }
+
+        skipBytes(buffer, 2);
+      }
+    } catch (BufferUnderflowException e) {
+      Log.e(TAG, "Malformed incoming packet");
```

Comment:
nit: Log.e(TAG, "Malformed incoming packet", e)?

Not sure if it's useful here. It may not be useful w/o the payload, but we should refrain from logging the payload itself on errors. Up to you.